src/execution.jl

```diff
@@ -118,6 +118,39 @@ for x in [:ExecutionEngine, :Interpreter, :JIT]
     end
 end
 
+Base.push!(engine::ExecutionEngine, mod::Module) = API.LLVMAddModule(engine.ref, mod.ref)
+
+# up to user to free the deleted module
+function Base.delete!(engine::ExecutionEngine, mod::Module)
+    out_ref = Ref{API.LLVMModuleRef}()
+    API.LLVMRemoveModule(engine.ref, mod.ref, out_ref, Ref{Cstring}()) # out string is not used
+    return Module(out_ref[])
+end
+
 run(engine::ExecutionEngine, f::Function, args::Vector{GenericValue}=GenericValue[]) =
     GenericValue(API.LLVMRunFunction(ref(engine), ref(f),
                                      Cuint(length(args)), ref.(args)))
+
+
+immutable ExecutionEngineFunctionSet
+    engine::ExecutionEngine
+end
+
+functions(engine::ExecutionEngine) = ExecutionEngineFunctionSet(engine)
+
+function __findfunction(functionset::ExecutionEngineFunctionSet, name::String)
```

Comment:
Why not immediately make this `get(functionset, name, nothing)` and use that in `haskey` and `getindex`?